<a href="https://colab.research.google.com/github/abishaalichan/Machine_Learning/blob/main/Deliquency_prediction(Tata_simulation).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**PROJECT - IDENTIFYING WHETHER AN ACCOUNT HOLDER CAN BE DELINQUENT OR NOT**

IMPORTING LIBRARIES

In [1]:
import numpy as np
import pandas as pd
import imblearn

READING FILE

In [2]:
df = pd.read_csv('/content/Delinquency_prediction_dataset.csv')

DISPLAYING FIVE RECORDS

In [3]:
df.head()

Customer_ID  Age    Income  Credit_Score  Credit_Utilization  \
0    CUST0001   56  165580.0         398.0            0.390502   
1    CUST0002   69  100999.0         493.0            0.312444   
2    CUST0003   46  188416.0         500.0            0.359930   
3    CUST0004   32  101672.0         413.0            0.371400   
4    CUST0005   60   38524.0         487.0            0.234716   

   Missed_Payments  Delinquent_Account  Loan_Balance  Debt_to_Income_Ratio  \
0                3                   0       16310.0              0.317396   
1                6                   1       17401.0              0.196093   
2                0                   0       13761.0              0.301655   
3                3                   0       88778.0              0.264794   
4                2                   0       13316.0              0.510583   

  Employment_Status  Account_Tenure Credit_Card_Type     Location Month_1  \
0               EMP              18          Student  Los Angeles    Late   
1     Self-employed               0         Standard      Phoenix  Missed   
2     Self-employed               1         Platinum      Chicago  Missed   
3        Unemployed              15         Platinum      Phoenix    Late   
4     Self-employed              11         Standard      Phoenix  Missed   

   Month_2 Month_3  Month_4  Month_5  Month_6  
0     Late  Missed     Late   Missed     Late  
1   Missed    Late   Missed  On-time  On-time  
2     Late    Late  On-time   Missed     Late  
3   Missed    Late   Missed     Late     Late  
4  On-time  Missed     Late     Late     Late

PERFORMING EDA

In [4]:
df.shape

(500, 19)

In [5]:
df.size

9500

In [6]:
df.isna().sum()

Customer_ID              0
Age                      0
Income                  39
Credit_Score             2
Credit_Utilization       0
Missed_Payments          0
Delinquent_Account       0
Loan_Balance            29
Debt_to_Income_Ratio     0
Employment_Status        0
Account_Tenure           0
Credit_Card_Type         0
Location                 0
Month_1                  0
Month_2                  0
Month_3                  0
Month_4                  0
Month_5                  0
Month_6                  0
dtype: int64

In [7]:
df.columns

Index(['Customer_ID', 'Age', 'Income', 'Credit_Score', 'Credit_Utilization',
       'Missed_Payments', 'Delinquent_Account', 'Loan_Balance',
       'Debt_to_Income_Ratio', 'Employment_Status', 'Account_Tenure',
       'Credit_Card_Type', 'Location', 'Month_1', 'Month_2', 'Month_3',
       'Month_4', 'Month_5', 'Month_6'],
      dtype='object')

In [8]:
for col in df.select_dtypes(include=['object']).columns:
    print(f"Unique values in {col}: {df[col].unique()}")
    print('-' * 40)

Unique values in Customer_ID: ['CUST0001' 'CUST0002' 'CUST0003' 'CUST0004' 'CUST0005' 'CUST0006'
 'CUST0007' 'CUST0008' 'CUST0009' 'CUST0010' 'CUST0011' 'CUST0012'
 'CUST0013' 'CUST0014' 'CUST0015' 'CUST0016' 'CUST0017' 'CUST0018'
 'CUST0019' 'CUST0020' 'CUST0021' 'CUST0022' 'CUST0023' 'CUST0024'
 'CUST0025' 'CUST0026' 'CUST0027' 'CUST0028' 'CUST0029' 'CUST0030'
 'CUST0031' 'CUST0032' 'CUST0033' 'CUST0034' 'CUST0035' 'CUST0036'
 'CUST0037' 'CUST0038' 'CUST0039' 'CUST0040' 'CUST0041' 'CUST0042'
 'CUST0043' 'CUST0044' 'CUST0045' 'CUST0046' 'CUST0047' 'CUST0048'
 'CUST0049' 'CUST0050' 'CUST0051' 'CUST0052' 'CUST0053' 'CUST0054'
 'CUST0055' 'CUST0056' 'CUST0057' 'CUST0058' 'CUST0059' 'CUST0060'
 'CUST0061' 'CUST0062' 'CUST0063' 'CUST0064' 'CUST0065' 'CUST0066'
 'CUST0067' 'CUST0068' 'CUST0069' 'CUST0070' 'CUST0071' 'CUST0072'
 'CUST0073' 'CUST0074' 'CUST0075' 'CUST0076' 'CUST0077' 'CUST0078'
 'CUST0079' 'CUST0080' 'CUST0081' 'CUST0082' 'CUST0083' 'CUST0084'
 'CUST0085' 'CUST0086' 'CUST0087

In [9]:
df.duplicated().sum()

np.int64(0)

FILLING MISSING VALUES BY IMPUTATION METHOD USING MEAN

In [10]:
df['Income'] = df['Income'].fillna(df['Income'].mean())
df['Credit_Score'] = df['Credit_Score'].fillna(df['Credit_Score'].mean())
df['Loan_Balance'] = df['Loan_Balance'].fillna(df['Loan_Balance'].mean())

In [11]:
df.isna().sum()

Customer_ID             0
Age                     0
Income                  0
Credit_Score            0
Credit_Utilization      0
Missed_Payments         0
Delinquent_Account      0
Loan_Balance            0
Debt_to_Income_Ratio    0
Employment_Status       0
Account_Tenure          0
Credit_Card_Type        0
Location                0
Month_1                 0
Month_2                 0
Month_3                 0
Month_4                 0
Month_5                 0
Month_6                 0
dtype: int64

REPLACING WRONG FORMAT DATA

In [12]:
df['Employment_Status'] = df['Employment_Status'].replace('employed', 'Employed')

In [13]:
df['Employment_Status'].unique()

array(['EMP', 'Self-employed', 'Unemployed', 'Employed', 'retired'],
      dtype=object)

In [14]:
df['Delinquent_Account'].value_counts()

Delinquent_Account
0    420
1     80
Name: count, dtype: int64

EXTRACTING KEY FEATURES NECESSARY FOR MODELING

In [15]:
df1 = df[['Age','Income','Credit_Score','Credit_Utilization','Missed_Payments','Debt_to_Income_Ratio','Delinquent_Account']]

SPLITTING INTO INPUT AND OUTPUT FEATURES

In [16]:
X=df1.drop('Delinquent_Account',axis=1)
y=df1['Delinquent_Account']

PERFORMING OVERSAMPLING USING SMOTE TECHNIQUE TO MAKE THE DATASET BALANCED

In [17]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X, y)

In [18]:
y_resampled.value_counts()

Delinquent_Account
0    420
1    420
Name: count, dtype: int64

SPLITTING INTO TRAINING AND TESTING DATA

In [19]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(X_resampled,y_resampled,test_size=0.2,random_state=42)

MODEL CREATION

In [20]:
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(random_state=42)
model.fit(x_train, y_train)

DecisionTreeClassifier(random_state=42)

PERFORMANCE EVALUATION

In [21]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

y_pred = model.predict(x_test)

print(f"Accuracy: {accuracy_score(y_test, y_pred)}")
print(f"Precision: {precision_score(y_test, y_pred)}")
print(f"Recall: {recall_score(y_test, y_pred)}")
print(f"F1 Score: {f1_score(y_test, y_pred)}")
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.7202380952380952
Precision: 0.6987951807228916
Recall: 0.725
F1 Score: 0.7116564417177914
Confusion Matrix:
[[63 25]
 [22 58]]


## Summary:

## Data Analysis Key Findings

*   The trained Decision Tree Classifier model achieved an accuracy of 0.7202 on the testing data.
*   The model demonstrated a precision of 0.6988 and a recall of 0.7250.
*   The F1 Score for the model is 0.7117.
*   The confusion matrix shows 63 True Negatives, 58 True Positives, 25 False Positives, and 22 False Negatives.

## Insights or Next Steps

*   The model shows a relatively balanced performance across precision and recall, indicated by the F1 score.
*   Further tuning of the Decision Tree model or exploring other algorithms might improve performance, especially regarding the balance between false positives and false negatives based on the business problem's requirements.
